code/modules/reagents/chemistry/equilibrium.dm

```diff
@@ -0,0 +1,293 @@
+	else
+		stack_trace("[holder.my_atom] attempted to determine FermiChem pH for '[reaction.type]' which had an invalid pH of [cached_pH] for set recipie pH vars. It's likely the recipe vars are wrong.")
+
+	//Calculate DeltaT (Deviation of T from optimal)
+	if(!reaction.is_cold_recipe)
+		if (cached_temp < reaction.optimal_temp && cached_temp >= reaction.required_temp)
+			deltaT = (((cached_temp - reaction.required_temp)**reaction.temp_exponent_factor)/((reaction.optimal_temp - reaction.required_temp)**reaction.temp_exponent_factor))
+		else if (cached_temp >= reaction.optimal_temp)
+			deltaT = 1
+		else //too hot
+			deltaT = 0
+			to_delete = TRUE
+			return
+	else
+		if (cached_temp > reaction.optimal_temp && cached_temp <= reaction.required_temp)
+			deltaT = (((cached_temp - reaction.required_temp)**reaction.temp_exponent_factor)/((reaction.optimal_temp - reaction.required_temp)**reaction.temp_exponent_factor))
+		else if (cached_temp <= reaction.optimal_temp)
+			deltaT = 1
+		else //Too cold
+			deltaT = 0
+			to_delete = TRUE
+			return
+
+	purity = deltapH//set purity equal to pH offset
+
+	//Then adjust purity of result with beaker reagent purity. 
+	purity *= reactant_purity(reaction)
+
+	//Now we calculate how much to add - this is normalised to the rate up limiter
+	var/delta_chem_factor = (reaction.rate_up_lim*deltaT)*delta_time//add/remove factor
+	var/total_step_added = 0
+	//keep limited
+	if(delta_chem_factor > target_vol)
+		delta_chem_factor = target_vol
+	else if (delta_chem_factor < CHEMICAL_VOLUME_MINIMUM)
+		delta_chem_factor = CHEMICAL_VOLUME_MINIMUM
+	//delta_chem_factor = round(delta_chem_factor, CHEMICAL_QUANTISATION_LEVEL) // Might not be needed - left here incase testmerge shows that it does. Remove before full commit.
+
+	//Calculate how much product to make and how much reactant to remove factors..
+	for(var/B in reaction.required_reagents)
+		holder.remove_reagent(B, ((delta_chem_factor/product_ratio) * reaction.required_reagents[B]), safety = TRUE)
+		//Apply pH changes
+		holder.adjust_specific_reagent_pH(B, ((delta_chem_factor/product_ratio) * reaction.required_reagents[B])*reaction.H_ion_release)
+
+	for(var/P in reaction.results)
+		//create the products
+		var/step_add = (delta_chem_factor/product_ratio) * reaction.results[P]
+		holder.add_reagent(P, step_add, null, cached_temp, purity, override_base_pH = TRUE)
+		//Apply pH changes
+		holder.adjust_specific_reagent_pH(P, step_add*reaction.H_ion_release)
+		reacted_vol += step_add
+		total_step_added += step_add
+
+	//#ifdef TESTING //Kept in so that people who want to write fermireactions can contact me with this log so I can help them
+	debug_world("Reaction vars: PreReacted:[reacted_vol] of [target_vol]. deltaT [deltaT], multiplier [multiplier], delta_chem_factor [delta_chem_factor] Pfactor [product_ratio], purity of [purity] from a deltapH of [deltapH]. DeltaTime: [delta_time]")
+	//#endif
```

Comment:
```suggestion
	#ifdef TESTING //Kept in so that people who want to write fermireactions can contact me with this log so I can help them
	debug_world("Reaction vars: PreReacted:[reacted_vol] of [target_vol]. deltaT [deltaT], multiplier [multiplier], delta_chem_factor [delta_chem_factor] Pfactor [product_ratio], purity of [purity] from a deltapH of [deltapH]. DeltaTime: [delta_time]")
	#endif
```
Back in.

Reply:
Fixed